IntegrationTests/tests_04_performance/test_01_resources/shared.swift

```diff
@@ -136,3 +136,133 @@ func withAutoReleasePool<T>(_ execute: () throws -> T) rethrows -> T {
+            // reduce allocations.
+            context.flush()
+        }
+
+        public func errorCaught(context: ChannelHandlerContext, error: Error) {
+            print("error: ", error)
+
+            // As we are not really interested getting notified on success or failure we just pass nil as promise to
+            // reduce allocations.
+            context.close(promise: nil)
+        }
+    }
+
+    private final class EchoHandlerClient: ChannelInboundHandler {
+        public typealias InboundIn = AddressedEnvelope<ByteBuffer>
+        public typealias OutboundOut = AddressedEnvelope<ByteBuffer>
+        private var numBytesOutstanding = 0
+        private var repetitionsRemaining : Int
+        
+        private let remoteAddressInitializer: () throws -> SocketAddress
+        
+        init(remoteAddressInitializer: @escaping () throws -> SocketAddress,
+             numberOfRepetitions: Int) {
+            self.remoteAddressInitializer = remoteAddressInitializer
+            self.repetitionsRemaining = numberOfRepetitions
+        }
+        
+        public func channelActive(context: ChannelHandlerContext) {
+            // Channel is available. It's time to send the message to the server to initialize the ping-pong sequence.
+            sendSomeDataIfDesiredOrClose(context: context)
+        }
+        
+        private func sendSomeDataIfDesiredOrClose(context: ChannelHandlerContext) {
+            if (repetitionsRemaining > 0) {
+                do {
+                    repetitionsRemaining -= 1
+                    
+                    // Get the server address.
+                    let remoteAddress = try self.remoteAddressInitializer()
+                    
+                    // Set the transmission data.
+                    let line = "Something to send there and back again."
+                    var buffer = context.channel.allocator.buffer(capacity: line.utf8.count)
+                    buffer.writeString(line)
+                    self.numBytesOutstanding += buffer.readableBytes
+                    
+                    // Forward the data.
+                    let envolope = AddressedEnvelope<ByteBuffer>(remoteAddress: remoteAddress, data: buffer)
+                    
+                    context.writeAndFlush(self.wrapOutboundOut(envolope), promise: nil)
+                    
+                } catch {
+                    print("Could not resolve remote address")
+                }
+            }
+            else {
+                // We're all done - hurrah!
+                context.close(promise: nil)
+            }
+        }
+        
+        public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+            let envelope = self.unwrapInboundIn(data)
+            let byteBuffer = envelope.data
+            
+            self.numBytesOutstanding -= byteBuffer.readableBytes
+            
+            if self.numBytesOutstanding <= 0 {
```

Comment:
We should probably remove this check: a datagram-oriented transport should expect one read per write, and not necessarily assume that all the bytes it wrote will come back.

Reply:
Done.